File: src/main/java/io/github/trainb0y/simplehud/config/gui/ConfigScreen.java
package io.github.trainb0y.simplehud.config.gui;

import dev.lambdaurora.spruceui.Position;
import dev.lambdaurora.spruceui.SpruceTexts;
import dev.lambdaurora.spruceui.background.DirtTexturedBackground;
import dev.lambdaurora.spruceui.option.SpruceBooleanOption;
import dev.lambdaurora.spruceui.option.SpruceIntegerInputOption;
import dev.lambdaurora.spruceui.option.SpruceSeparatorOption;
import dev.lambdaurora.spruceui.option.SpruceSimpleActionOption;
import dev.lambdaurora.spruceui.option.SpruceStringOption;
import dev.lambdaurora.spruceui.option.SpruceToggleBooleanOption;
import dev.lambdaurora.spruceui.screen.SpruceScreen;
import dev.lambdaurora.spruceui.widget.SpruceButtonWidget;
import dev.lambdaurora.spruceui.widget.container.SpruceOptionListWidget;
import io.github.trainb0y.simplehud.config.Config;
import net.minecraft.client.MinecraftClient;
import net.minecraft.client.gui.screen.Screen;
import net.minecraft.client.util.math.MatrixStack;
import net.minecraft.text.Text;
import org.jetbrains.annotations.Nullable;

import java.util.List;

/**
 * Primary configuration screen
 */
public class ConfigScreen extends SpruceScreen {
    private final Screen parent;
    private final MinecraftClient client = MinecraftClient.getInstance();
    private SpruceOptionListWidget optionList;


    public ConfigScreen(@Nullable Screen parent) {
        super(Text.translatable("config.simplehud.title"));
        this.parent = parent;
    }

    @Override
    public void renderTitle(MatrixStack matrices, int mouseX, int mouseY, float delta) {
        drawCenteredText(matrices, this.textRenderer, this.title, this.width / 2, 8, 0xFFFFFF);
    }


    @Override
    protected void init() {
        super.init();

        this.optionList = new SpruceOptionListWidget(Position.of(0, 22), this.width, this.height - 36 - 22);
        this.optionList.setBackground(DirtTexturedBackground.DARKENED);
        this.optionList.addOptionEntry(new SpruceBooleanOption("config.simplehud.option.enabled",
                () -> Config.hudEnabled,
                value -> Config.hudEnabled = value,
                Text.translatable("config.simplehud.option.enabled.tooltip")
        ), SpruceSimpleActionOption.of("config.simplehud.editpositions",
                btn -> {
                    client.setScreen(new PositionScreen(this, Config.elements.stream().filter(element -> element.enabled).toList()));
                },
                Text.translatable("config.simplehud.editposition.tooltip")
        ));

        Config.elements.forEach((element) -> {
            this.optionList.addSingleOptionEntry(new SpruceSeparatorOption(element.getKey() + ".name", true, Text.translatable(element.getKey() + ".tooltip")));
            this.optionList.addOptionEntry(new SpruceToggleBooleanOption("config.simplehud.enabled",
                            () -> element.enabled,
                            value -> element.enabled = value,
                            null
                    ),
                    SpruceSimpleActionOption.of("config.simplehud.editposition",
                            btn -> {
                                client.setScreen(new PositionScreen(this, List.of(element)));
                            },
                            Text.translatable("config.simplehud.editposition.tooltip")
                    ));
            this.optionList.addSingleOptionEntry(new SpruceStringOption("config.simplehud.override",
                    () -> element.override,
                    value -> {
                        if (value == "") element.override = null;
                        else element.override = value;
                    },
                    null,
                    Text.translatable("config.simplehud.override.tooltip")
            ));

            this.optionList.addOptionEntry(new SpruceIntegerInputOption("config.simplehud.editposition.x",
                    () -> element.x,
                    value -> element.x = value,
                    Text.translatable("config.simplehud.editposition.x.tooltip")
            ), new SpruceIntegerInputOption("config.simplehud.editposition.y",
                    () -> element.y,
                    value -> element.y = value,
                    Text.translatable("config.simplehud.editposition.y.tooltip")
            ));
        });

        this.addDrawableChild(this.optionList);

        this.addDrawableChild(new SpruceButtonWidget(Position.of(this, this.width / 2 - 155 + 160, this.height - 29), 150, 20, SpruceTexts.GUI_DONE,
                (btn) -> {
                    Config.saveConfig();
                    this.client.setScreen(this.parent);
                }).asVanilla());

        this.addDrawableChild(new SpruceButtonWidget(Position.of(this, this.width / 2 - 155, this.height - 29), 150, 20, SpruceTexts.CONTROLS_RESET,
                (btn) -> {
                    Config.applyDefaultConfig();
                    Config.saveConfig();
                    this.client.setScreen(new ConfigScreen(this.parent)); // values don't automatically update, so...
                }).asVanilla());
    }

}

File: src/main/kotlin/io/github/trainb0y/fabrihud/config/gui/ConfigScreen.kt
package io.github.trainb0y.fabrihud.config.gui

import dev.isxander.yacl.api.Binding
import dev.isxander.yacl.api.ButtonOption
import dev.isxander.yacl.api.ConfigCategory
import dev.isxander.yacl.api.Option
import dev.isxander.yacl.api.OptionGroup
import io.github.trainb0y.fabrihud.config.Config
import io.github.trainb0y.fabrihud.elements.Element
import io.github.trainb0y.fabrihud.elements.TextElement
import dev.isxander.yacl.api.YetAnotherConfigLib
import dev.isxander.yacl.gui.YACLScreen
import dev.isxander.yacl.gui.controllers.ActionController
import dev.isxander.yacl.gui.controllers.TickBoxController
import dev.isxander.yacl.gui.controllers.string.StringController
import dev.isxander.yacl.gui.controllers.slider.IntegerSliderController
import net.minecraft.client.MinecraftClient
import net.minecraft.client.gui.screen.Screen
import net.minecraft.text.Text


fun openConfigScreen(parent: Screen?): Screen =
	// Welcome to Builder Hell:tm:
	// We hope you brought popcorn!

	// tbh the SpruceUI config screen was a lot cleaner,
	// but it was kind of ugly, and FabriZoom needed some YACL features,
	// and I'd rather have consistency between mods.

	YetAnotherConfigLib.createBuilder()
		.title(Text.translatable("config.fabrihud.global"))
		.category(ConfigCategory.createBuilder()
			.name(Text.translatable("config.fabrihud.primary"))
			.option(Option.createBuilder(Boolean::class.java)
				.name(Text.translatable("config.fabrihud.option.enabled"))
				.tooltip(Text.translatable("config.fabrihud.option.enabled.tooltip"))
				.binding(
					Binding.generic(
						Config.hudEnabled,
						{ Config.hudEnabled },
						{value -> Config.hudEnabled = value}
					)
				)
				.controller { option ->
					TickBoxController(option)
				}
				.build()
			)
			.option(ButtonOption.createBuilder()
				.name(Text.translatable("config.fabrihud.editpositions"))
				.tooltip(Text.translatable("config.fabrihud.editpositions.tooltip"))
				.action { screen: YACLScreen?, _: ButtonOption? ->
					MinecraftClient.getInstance().setScreen(
						PositionScreen(
							screen as Screen,
							Config.elements.stream().filter { element: Element -> element.enabled }.toList()
						)
					)
				}.controller { option: ButtonOption? ->
					ActionController(option)
				}
				.build()
			)
			.option(ButtonOption.createBuilder()
				.name(Text.translatable("config.fabrihud.reset"))
				.tooltip(Text.translatable("config.fabrihud.reset.tooltip"))
				.action { screen: YACLScreen?, _: ButtonOption? ->
					Config.applyDefaultConfig()
					Config.saveConfig()
					screen?.close()

				}.controller { option: ButtonOption? ->
					ActionController(option)
				}
				.build()
			)
			.build()
		)
		.category(ConfigCategory.createBuilder()
			.name(Text.translatable("config.fabrihud.elements"))
			.also { category ->
				Config.elements.forEach { element ->
					category.group(OptionGroup.createBuilder()
						.name(Text.translatable(element.key + ".name"))
						.tooltip(Text.translatable(element.key + ".tooltip"))
						.option(Option.createBuilder(Boolean::class.java)
							.name(Text.translatable("config.fabrihud.enabled"))
							.tooltip(Text.translatable("config.fabrihud.enabled.tooltip"))
							.binding(
								Binding.generic(
									element.enabled,
									{ element.enabled },
									{value -> element.enabled = value}
								)
							)
							.controller { option ->
								TickBoxController(option)
							}
							.build()
						)
						.option(ButtonOption.createBuilder()
							.name(Text.translatable("config.fabrihud.editposition"))
							.tooltip(Text.translatable("config.fabrihud.editposition.tooltip"))
							.action { screen: YACLScreen?, _: ButtonOption? ->
								MinecraftClient.getInstance().setScreen(
									PositionScreen(
										screen as Screen,
										listOf(element)
									)
								)
							}.controller { option: ButtonOption? ->
								ActionController(option)
							}
							.build()
						)
						.also {group ->
							if (element !is TextElement) return@also
							group.option(Option.createBuilder(String::class.java)
								.name(Text.translatable("config.fabrihud.override"))
								.tooltip(Text.translatable("config.fabrihud.override.tooltip"))
								.binding(
									Binding.generic(
										element.override ?: "",
										{ element.override ?: ""},
										{value -> element.override = value}
									)
								)
								.controller { option ->
									StringController(option)
								}
								.build()
							)
							group.option(Option.createBuilder(Boolean::class.java)
								.name(Text.translatable("config.fabrihud.shadow"))
								.tooltip(Text.translatable("config.fabrihud.shadow.tooltip"))
								.binding(
									Binding.generic(
										element.shadow,
										{ element.shadow },
										{value -> element.shadow = value}
									)
								)
								.controller { option ->
									TickBoxController(option)
								}
								.build()
							)
							group.option(Option.createBuilder(Boolean::class.java)
								.name(Text.translatable("config.fabrihud.background"))
								.tooltip(Text.translatable("config.fabrihud.background.tooltip"))
								.binding(
									Binding.generic(
										element.background,
										{ element.background },
										{value -> element.background = value}
									)
								)
								.controller { option ->
									TickBoxController(option)
								}
								.build()
							)
						}
						.option(Option.createBuilder(Int::class.java)
							.name(Text.translatable("config.fabrihud.editposition.x"))
							.tooltip(Text.translatable("config.fabrihud.editposition.x.tooltip"))
							.binding(
								Binding.generic(
									element.x,
									{ element.x },
									{value -> element.x = value}
								)
							)
							.controller { option ->
								IntegerSliderController(
									option,
									0,
									MinecraftClient.getInstance().window.width,
									1
								)
							}
							.build()
						)
						.option(Option.createBuilder(Int::class.java)
							.name(Text.translatable("config.fabrihud.editposition.y"))
							.tooltip(Text.translatable("config.fabrihud.editposition.y.tooltip"))
							.binding(
								Binding.generic(
									element.y,
									{ element.y },
									{value -> element.y = value}
								)
							)
							.controller { option ->
								IntegerSliderController(
									option,
									0,
									MinecraftClient.getInstance().window.height,
									1
								)
							}
							.build()
						)
						.build()
					)
				}
			}
			.build()
		)
		.save(Config::saveConfig)
		.build()
		.generateScreen(parent)